contributors/design-proposals/auth/kubectl-exec-plugins.md

```diff
@@ -0,0 +1,264 @@
+    "invalidCredentials": true
+}'
+```
+
+### Caching
+
+kubectl repeatedly [re-initializes transports](https://github.com/kubernetes/kubernetes/issues/37876)
+while client-go transports are long lived over many requests. As a result naive auth
+provider implementations that re-request credentials on every request have historically
+been slow.
+
+Plugins will be called on client-go initialization, and again when the API server returns
+a 401 HTTP status code indicating expired credentials. Plugins can indicate their credentials
+explicit expiry using the `Expiry` field on the returned `ExecCredentials` object, otherwise
+credentials will be cached hroughout the lifetime of a program.
+
+## Kubeconfig changes
+
+The current `AuthProviderConfig` uses `map[string]string` for configuration, which
+makes it hard to express things like a list of arguments or list key/value environment
+variables. As such, `AuthInfo` should add another field which expresses the `exec`
+config. This has the benefit of a more natural structure, but the trade-off of not being
+compatible with the existing `kubectl config set-credentials` implementation.
+
+```go
+// AuthInfo contains information that describes identity information.  This is use to tell the kubernetes cluster who you are.
+type AuthInfo struct {
+    // Existing fields ...
+
+    // Exec is a command to execute which returns credentials to the transport to use.
+    // +optional
+    Exec *ExecAuthProviderConfig `json:"exec,omitempty"`
+
+    // ...
+}
+
+type ExecAuthProviderConfig struct {
+    Command string   `json:"command"`
+    Args    []string `json:"args"`
+    // Env defines additional environment variables to expose to the process. These
+    // are unioned with the host's environment, as well as variables client-go uses
+    // to pass argument to the plugin.
+    Env []ExecEnvVar `json:"env"`
+
+    // Prefered input version of the ExecInfo. The returned ExecCredentials MUST use
+    // the same encoding version as the input.
+    APIVersion string `json:"apiVersion,omitempty"`
+
+    // TODO: JSONPath options for filtering output.
+}
+
+type ExecEnvVar struct {
+    Name  string `json:"name"`
+    Value string `json:"value"`
+
+    // TODO: Load env vars from files or from other envs?
+}
+```
+
+This would allow a user block of a kubeconfig to declare the following:
+
+```yaml
+users:
+- name: mmosley
+  user:
+    exec:
+      apiVersion: "client.authentication.k8s.io/v1alpha1"
+      command: /bin/kubectl-login
+      args: ["hello", "world"]
+```
+
+The AWS authenticator, modified to return structured output, would become:
+
+```
+users:
+- name: kubernetes-admin
+  user:
+    exec:
+      apiVersion: "client.authentication.k8s.io/v1alpha1"
+      command: heptio-authenticator-aws
+      # CLUSTER_ID and ROLE_ARN should be replaced with actual desired values.
+      args: ["token", "-i", "(CLUSTER_ID)", "-r", "(ROLE_ARN)"]
+```
+
+## TLS client certificate support
```

Comment:
Not necessarily a problem with this proposal, but I wanted to mention the other x509 use case I've had which was to use certs/keys that are on a PKCS#11 token/smart card. In this case the auth provider wouldn't be able to pass back private keys, it would be like a `crypto.Signer` that supported signing/decryption operations using the non-exportable key.

Reply:
at least initially, that seems like it would be easier to deal with as a local proxy using something like https://github.com/square/ghostunnel#hsmpkcs11-support (similar to how a bastion setup could work)

Reply:
I would have expected integrations with things like yubikeys to sign short lived tokens instead of integrating into the TLS handshake.